src/gui/qgshelp.cpp

```diff
@@ -0,0 +1,169 @@
+/***************************************************************************
+  qgshelp.cpp
+  --------------------------------------
+  Date                 : December 2016
+  Copyright            : (C) 2016 by Alexander Bruy
+  Email                : alexander dot bruy at gmail dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgshelp.h"
+
+#include <QSettings>
+#include <QLocale>
+#include <QUrl>
+#include <QFileInfo>
+#include <QTcpSocket>
+#include <QDesktopServices>
+#include <QNetworkProxy>
+#include <QNetworkProxyFactory>
+
+#include "qgis.h"
+#include "qgsapplication.h"
+
+void QgsHelp::openHelp( const QString& key )
+{
+  QDesktopServices::openUrl( QgsHelp::helpUrl( key ) );
+}
+
+QUrl QgsHelp::helpUrl( const QString& key )
+{
+  QUrl helpNotFound = QUrl::fromLocalFile( QgsApplication::pkgDataPath() + "/doc/nohelp.html" );
+
+  QSettings settings;
+  QStringList paths = settings.value( QStringLiteral( "help/helpSearchPath" ) ).toStringList();
+  if ( paths.isEmpty() )
+  {
+    return helpNotFound;
+  }
+
+  QString qgisLocale;
+  bool overrideLocale = settings.value( QStringLiteral( "locale/overrideFlag" ), false ).toBool();
+  if ( overrideLocale )
+  {
+    qgisLocale = settings.value( QStringLiteral( "locale/userLocale" ), QString() ).toString();
+  }
+  else
+  {
+    qgisLocale = QLocale::system().name().left( 2 );
+  }
+
+  QString qgisVersion;
+  if ( Qgis::QGIS_VERSION_INT / 100 % 100 == 99 )
+  {
+    qgisVersion = QStringLiteral( "testing" );
+    qgisLocale = QStringLiteral( "en" );
+  }
+  else
+  {
+    qgisVersion = QStringLiteral( "%1.%2" ).arg( Qgis::QGIS_VERSION_INT / 10000 ).arg( Qgis::QGIS_VERSION_INT / 100 % 100 );
+  }
+
+  QString suffix = QStringLiteral( "%1/%2/docs/user_manual/%3" ).arg( qgisVersion ).arg( qgisLocale ).arg( key );
```

Comment:
Not sure I understand correctly. Original goal was to provide some way to use QGIS docs also as context help/Processing algorithm help source, that's why all keys handled in the same way.

Are you talking about implementing some logic to distinguish QGIS related keys from all other keys and using different help sources for each key type?